lab_4/main.py

```diff
@@ -66,41 +107,108 @@ class NGramTextGenerator:
+                if n_gram_index == len(possible_n_grams) - 1 and self._used_n_grams:
+                    self._used_n_grams.clear()
+                    n_gram_index = 0
+            if n_gram_index == len(possible_n_grams) - 1 and not self._used_n_grams:
+                break
+        return possible_n_grams[0][-1]
 
     def _generate_word(self, context: tuple, word_max_length=15) -> tuple:
         """
         Generates full word for the context given.
         """
-        pass
+        if not isinstance(context, tuple) or not isinstance(word_max_length, int):
+            return ()
+        if len(context) >= word_max_length:
+            return_context = []
+            for letter_id in context:
+                return_context.append(letter_id)
+            return_context.append(self.profile.storage.get_special_token_id())
+            return tuple(return_context)
+
+        word = []
+        word.extend(context)
+        while len(word) < word_max_length:
+            word.append(self._generate_letter(tuple(context)))
+            context = [word[-1]]
+            if word[-1] == self.profile.storage.get_special_token_id():
+                return tuple(word)
+            if len(word) == word_max_length:
+                word.append(self.profile.storage.get_special_token_id())
+                return tuple(word)
+        return ()
 
     def generate_sentence(self, context: tuple, word_limit: int) -> tuple:
         """
         Generates full sentence with fixed number of words given.
         """
-        pass
+        if not isinstance(context, tuple) or not isinstance(word_limit, int) or word_limit <= 0:
+            return ()
+        sentence = []
+        word_counter = 0
+        sentence.append(self._generate_word(context))
+        while len(sentence) < word_limit:
+            context = [sentence[word_counter][-1]]
+            sentence.append(self._generate_word(tuple(context)))
+            word_counter += 1
+        return tuple(sentence)
 
     def generate_decoded_sentence(self, context: tuple, word_limit: int) -> str:
         """
         Generates full sentence and decodes it
         """
-        pass
+        sentence = self.generate_sentence(context, word_limit)
+        decoded_sentence = []
+        for word in sentence:
+            decoded_word = []
+            for letter_id in word:
+                decoded_word.append(self.profile.storage.get_element(letter_id))
+            decoded_sentence.append(decoded_word)
+        return translate_sentence_to_plain_text(tuple(decoded_sentence))
 
 
 # 6
 def translate_sentence_to_plain_text(decoded_corpus: tuple) -> str:
     """
     Converts decoded sentence into the string sequence
     """
-    pass
+    if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+        return ''
+    sentence = []
+    for word in decoded_corpus:
+        sentence.extend(word)
+    if sentence[0] == '_':
+        sentence[0] = ''
+    if sentence[-1] == '_':
```

Comment:
Could be combined into `if-elif`

Reply:
We cannot combine these because we need to check both the beginning and the end of a sentence